pkg/descheduler/evictions/evictions.go

```diff
@@ -121,13 +478,35 @@ type EvictOptions struct {
 // EvictPod evicts a pod while exercising eviction limits.
 // Returns true when the pod is evicted on the server side.
 func (pe *PodEvictor) EvictPod(ctx context.Context, pod *v1.Pod, opts EvictOptions) error {
+	if len(pod.UID) == 0 {
+		klog.InfoS("Ignoring pod eviction due to missing UID", "pod", pod)
+		return fmt.Errorf("Pod %v is missing UID", klog.KObj(pod))
+	}
+
+	if pe.featureGates.Enabled(features.EvictionsInBackground) {
+		if pod.Annotations != nil {
+			// eviction in background requested
+			if _, exists := pod.Annotations[EvictionRequestAnnotationKey]; exists {
+				exists, err := pe.erCache.hasPod(pod)
+				if err != nil {
+					return fmt.Errorf("unable to check whether a pod exists in the cache of eviction requests: %v", err)
+				}
+				if exists {
+					klog.V(3).InfoS("Eviction in background already requested (ignoring)", "pod", klog.KObj(pod))
+					return nil
+				}
+			}
+		}
+	}
+
 	pe.mu.Lock()
 	defer pe.mu.Unlock()
+
 	var span trace.Span
 	ctx, span = tracing.Tracer().Start(ctx, "EvictPod", trace.WithAttributes(attribute.String("podName", pod.Name), attribute.String("podNamespace", pod.Namespace), attribute.String("reason", opts.Reason), attribute.String("operation", tracing.EvictOperation)))
 	defer span.End()
 
-	if pe.maxPodsToEvictTotal != nil && pe.totalPodCount+1 > *pe.maxPodsToEvictTotal {
+	if pe.maxPodsToEvictTotal != nil && pe.totalPodCount+pe.erCache.evictionRequestsTotal()+1 > *pe.maxPodsToEvictTotal {
```

Comment:
Shouldn't this be gated?

Reply:
It probably should to avoid cases where `pe.erCache.evictionRequestsTotal()` misbehaves. In the current implementation `evictionRequestsCache.requestsTotal` will always be 0 when the feature gate is not enabled.

Reply:
Done